src/gDP.cpp

```diff
@@ -632,7 +632,7 @@ void gDPLoadTLUT( u32 tile, u32 uls, u32 ult, u32 lrs, u32 lrt )
 	u16 *dest = (u16*)&TMEM[gDP.tiles[tile].tmem];
 
 	int i = 0;
-	while (i < count) {
+	while (i < count && pal < 16) {
```

Comment:
Also, I'm not entirely sure what correct behavior should be. Should we wrap around instead of exiting the loop?

Reply:
This is impossible situation. pal can't be >= 16.
Palette has either 16 or 256 elements (count).
For 256 elements palette initial pal index **must be** 0.
Or if pal is not zero, count **must be** 16
If it is not so, something bad happened before. This fix masks the real problem. Speaking of which: I have no problems with your save on Windows. I tried to load it on Ubuntu and first got that impossible situation and then, after memory corruption, got assert in F3DEX2.cpp, line 68. I suppose this is an issue in mupen64plus core: such problem can happen when at some moment gDPSetTile is called, but corresponding gDPSetTileSize does not, thus palette count calculated incorrectly.

I can't check, does the older mupen64plus core has this issue, because your save can't be loaded with older core. Most likely we have to add a workaround for  this problem on plugin's side. Your fix works, but it is wrong, because it spoils palettes with index above pal. I suggest to add clearly visible workaround, like this:
```
	u16 count = (u16)((gDP.tiles[tile].lrs - gDP.tiles[tile].uls + 1) * (gDP.tiles[tile].lrt - gDP.tiles[tile].ult + 1));
	u32 address = gDP.textureImage.address + gDP.tiles[tile].ult * gDP.textureImage.bpl + (gDP.tiles[tile].uls << gDP.textureImage.size >> 1);
	u16 pal = (u16)((gDP.tiles[tile].tmem - 256) >> 4);
	u16 *dest = (u16*)&TMEM[gDP.tiles[tile].tmem];
#if defined(MUPENPLUSAPI) && !defined(OS_WINDOWS)
    // Workaround for mupen64plus error
    if (pal != 0)
        count = 16;
#endif
```

Reply:
@fzurita did you create that save state? Didn't retroben first report this issue? Is it possible that some cheat has corrupted the memory?

Reply:
It was created by retroben.  I don't think a cheat should cause corruption in the display list itself. I guess @retrobenny could always create an in-game save and get to the same spot from the beginning to if the issue happens again.

Reply:
Whatever the issue might be, I don't think it's specific to mupen64plus, another person reported the same issue with PJ64: https://github.com/gonetz/GLideN64/issues/1250

Reply:
@gonetz Also, my fix is intended to protect memory. If pal can really be greater than 16 then paletteCRC16 must be enlarged to allow for a bigger size.

Reply:
> If pal can really be greater than 16

It **must not ** be greater than 16. By N64 design.

Reply:
> another person reported the same issue with PJ64

another person could use cheats as well.

For me the problem is platform-specific.

Reply:
> For me the problem is platform-specific.

Just because it didn't crash doesn't mean the problem doesn't exist. When dealing with out-of-bounds memory issues each computer might behave a bit differently. You might have gotten lucky on your Windows box, but someone else might not. There isn't really any platform specific emulation code in mupen64plus or GLideN64

Reply:
Take this issue for instance (Paper Mario crash):
https://github.com/gonetz/GLideN64/issues/1539

Same thing (bogus pointer this time), it crashed in Linux, but not Windows.

Anyway I'm not saying there is necessarily something wrong with GLideN64, it could be a bad savestate corrupted by a cheat, I just don't think the issue is specific to mupen64plus or Linux.

Reply:
@gonetz Maybe instead of performing this check ```pal < 16``` we can do this instead?

```u16 pal = (u16)(((gDP.tiles[tile].tmem - 256) >> 4) & 0xF);```

I just added a 0xF to mask the value generated from the tmem parameter.

Reply:
You did not read carefully my explanation of the problem. The problem is not with the pallette index pal, but with tile size, which defines pallette size  (count). If pal is not zero, count **must be** 16. My fix insures that. I agree that it is too strict to apply that fix only for mupen64plus and Linux. This condition must always be satisfied. Comment, why it is added is necessary anyway.